### PR TITLE
Update LargestSubstringWithoutCharacters.java

### PR DESCRIPTION
**Pull request by BoredomCrusher**
creating pull request for code review

### CODE REVIEW

File: String/LargestSubstringWithoutCharacters.java [hunk 1]
```diff
@@ -1 +1,62 @@
...
+// code is written to solve leetcode problem 3
+public class LargestSubstringWithoutCharacters {
+    public static void main(String[] args) {
+        LargestSubstringWithoutCharacters largestString = new LargestSubstringWithoutCharacters();
+
+        Boolean testCase1 = largestString.lengthOfLongestSubstring("abcabcbb") == 3;
+        Boolean testCase2 = largestString.lengthOfLongestSubstring("bbbbb") == 1;
+        Boolean testCase3 = largestString.lengthOfLongestSubstring("pwwkew") == 3;
+        Boolean testCase4 = largestString.lengthOfLongestSubstring("dvdf") == 3;
+        Boolean testCase5 = largestString.lengthOfLongestSubstring("abba") == 2;
```
Comment (tshur) [inline String/LargestSubstringWithoutCharacters.java:14]: It's awesome that you are writing tests. Two nit comments:

1. If you are doing tests like this, you might as well learn how to do official tests in java (i.e., write a separate test file).
2. You missed some edge cases that would be good to have checked: `""` (empty string) and `"abc"` (the whole string is one longest substring).

File: String/LargestSubstringWithoutCharacters.java [hunk 2]
```diff
@@ -1 +1,62 @@
+package String;
+
+import java.util.HashMap;
+
+// code is written to solve leetcode problem 3
+public class LargestSubstringWithoutCharacters {
```
Comment (tshur) [inline String/LargestSubstringWithoutCharacters.java:6]: nit: naming is confusing "WithoutCharacters" -- I had to look up the leetcode problem to see its "Without*Repeating*Characters"

File: String/LargestSubstringWithoutCharacters.java [hunk 3]
```diff
@@ -1 +1,62 @@
...
+        if (s.length() <= 1) {
+            return s.length();
+        }
+
+        HashMap<Character, Boolean> map = new HashMap<>();
+        String tempString = "";
+        int finalStringLength = 0;
+
+        for (int i = 0; i < s.length(); i++) {
+            // if next character has already been added to the hashmap
```
Comment (tshur) [inline String/LargestSubstringWithoutCharacters.java:33]: Do you want me to leave style guide critiques? For example, at my company all comments are written as a sentence (capital + period, ideally an independent clause):

`// If next character has already been added to the hashmap.`

File: String/LargestSubstringWithoutCharacters.java [hunk 4]
```diff
@@ -1 +1,62 @@
...
+    public int lengthOfLongestSubstring(String s) {
+        if (s.length() <= 1) {
+            return s.length();
+        }
+
+        HashMap<Character, Boolean> map = new HashMap<>();
+        String tempString = "";
+        int finalStringLength = 0;
+
+        for (int i = 0; i < s.length(); i++) {
```
Comment (tshur) [inline String/LargestSubstringWithoutCharacters.java:32]: I noticed that you don't use the variable `i` *except* to get the current `charAt(...)`. Seems like a great candidate for a `for-each` loop (https://www.geeksforgeeks.org/for-each-loop-in-java/) -- also below with your `j` loop. This makes it much more obvious that you are going character-by-character (as opposed to jumping around indices or reading i+1 etc)

File: String/LargestSubstringWithoutCharacters.java [hunk 5]
```diff
@@ -1 +1,62 @@
...
+            // if next character has already been added to the hashmap
+            if (map.get(s.charAt(i)) != null) {
+                // get string after first instance of repeated character
+                tempString = tempString.split("" + s.charAt(i), 2)[1];
+
+                // reset hashmap and fill with characters of current string
+                // -- I have no idea why just removing the current character doesn't work
+                // -- but somehow, that doesn't work, and this works
+                // -- for every testcase on leetcode
```
Comment (tshur) [inline String/LargestSubstringWithoutCharacters.java:41]: This does not seem like a case that needs a bulleted list (my guess for what you are trying to do. Also, the lines are different lengths. How about just an 80char max-width paragraph?

```
// Reset HashMap and fill with characters of current string. I have no
// idea why just removing the current character doesn't work, but
// somehow, that doesn't work. This works for every testcase on leetcode.
```

File: String/LargestSubstringWithoutCharacters.java [hunk 6]
```diff
@@ -1 +1,62 @@
...
+
+        HashMap<Character, Boolean> map = new HashMap<>();
+        String tempString = "";
+        int finalStringLength = 0;
+
+        for (int i = 0; i < s.length(); i++) {
+            // if next character has already been added to the hashmap
+            if (map.get(s.charAt(i)) != null) {
+                // get string after first instance of repeated character
+                tempString = tempString.split("" + s.charAt(i), 2)[1];
```
Comment (tshur) [inline String/LargestSubstringWithoutCharacters.java:36]: This feels a little bit brittle / complicated. To me, this is not "obvious that it's working". Also, I dislike the "" + char hack, but I guess `new String(char)` is not great either. I thought `.substring(...indexOf(...))` might be better, but then you have to worry if the character isn't in the string.

File: String/LargestSubstringWithoutCharacters.java [hunk 7]
```diff
@@ -1 +1,62 @@
...
+
+                // reset hashmap and fill with characters of current string
+                // -- I have no idea why just removing the current character doesn't work
+                // -- but somehow, that doesn't work, and this works
```
Comment (tshur) [inline String/LargestSubstringWithoutCharacters.java:40]: It doesn't work to simply remove a character from the hashmap since you may actually have to remove multiple characters in some cases.

Example:
"abcb"

Suppose you find "abc" and then encounter "b".
If you try to remove "b" from the hashmap, then your hashmap will still contain "a". However, this is a mistake. By re-creating the hashmap, you are actually doing things correctly, since you are cutting "a" out of your tempString.split(...).

However, rebuilding this hashmap is not super efficient.

File: String/LargestSubstringWithoutCharacters.java [hunk 8]
```diff
@@ -1 +1,62 @@
...
+                // -- for every testcase on leetcode
+                map = new HashMap<>();
+                for (int j = 0; j < tempString.length(); j++) {
+                    map.put(tempString.charAt(j), true);
+                }
+            }
+
+            // add current character to hashmap
```
Comment (tshur) [inline String/LargestSubstringWithoutCharacters.java:48]: To me, this is a redundant comment.

File: String/LargestSubstringWithoutCharacters.java [hunk 9]
```diff
@@ -1 +1,62 @@
...
+        Boolean testCase5 = largestString.lengthOfLongestSubstring("abba") == 2;
+
+        System.out.println("testCase1: " + testCase1);
+        System.out.println("testCase2: " + testCase2);
+        System.out.println("testCase3: " + testCase3);
+        System.out.println("testCase4: " + testCase4);
+        System.out.println("testCase4: " + testCase5);
+    }
+
+    public int lengthOfLongestSubstring(String s) {
```
Comment (tshur) [inline String/LargestSubstringWithoutCharacters.java:23]: What is the runtime complexity of this function as it is written?

**Review by tshur — APPROVED** [review]
